src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -266,22 +347,53 @@ public void Save(ModelSaveContext ctx)
             _host.AssertValue(ctx);
             ctx.CheckAtModel();
             ctx.SetVersionInfo(GetVersionInfo());
+            ctx.Writer.Write(IsFrozen ? 1 : 0);
+
             // *** Binary format ***
+            // int: indicator for frozen models
             // stream: tensorFlow model.
             // int: number of input columns
             // for each input column
             //   int: id of int column name
             // int: number of output columns
             // for each output column
             //   int: id of output column name
+            if (IsFrozen)
+            {
+                var buffer = new TFBuffer();
+                Session.Graph.ToGraphDef(buffer);
 
-            var buffer = new TFBuffer();
-            Session.Graph.ToGraphDef(buffer);
-
-            ctx.SaveBinaryStream("TFModel", w =>
+                ctx.SaveBinaryStream("TFModel", w =>
+                {
+                    w.WriteByteArray(buffer.ToArray());
+                });
+            }
+            else
             {
-                w.WriteByteArray(buffer.ToArray());
-            });
+                var savedModel = new Dictionary<string, byte[]>();
+                if (!string.IsNullOrEmpty(ModelPath))
+                {
+                    string[] modelFilePaths = Directory.GetFiles(ModelPath, "*", SearchOption.AllDirectories);
+                    foreach (var path in modelFilePaths)
+                    {
+                        var relativePath = path.Remove(0, ModelPath.Length).Trim(Path.DirectorySeparatorChar);
+                        savedModel[relativePath] = File.ReadAllBytes(path);
```

Comment:
>savedModel [](start = 24, length = 10)

Rather than doing the ReadAllBytes here and loading every file into memory, can you instead open it in the SaveBinaryStream callback and do CopyTo using the backing stream?  That will avoid loading all the files into memory at once.

Reply:
Looks like you now moved that into the constructor which is worse (and presumably incorrect since it's snapshotting the files when this transform is created, instead of when save is called).

---
In reply to: [217455187](https://github.com/dotnet/machinelearning/pull/853#discussion_r217455187) [](ancestors = 217455187)

Reply:
Actually, I commented to move it to constructor because files will be cached when loading from temp directory anyway. It makes logic bit simpler as well. 
Also, moving the code into constructor will remove the folder dependency on save i.e. actually folder is not needed when saving the model. I think there might be situations when `ModelPath`  wont be available during save and load process e.g. when saving already trained ML.Net model.

---
In reply to: [217520354](https://github.com/dotnet/machinelearning/pull/853#discussion_r217520354) [](ancestors = 217520354,217455187)

Reply:
No problem, I am fine going back to the previous way.

---
In reply to: [217552071](https://github.com/dotnet/machinelearning/pull/853#discussion_r217552071) [](ancestors = 217552071,217526355,217520354,217455187)